velox/expression/VectorUdfTypeSystem.h

```diff
@@ -340,6 +341,107 @@ struct VectorWriter<Map<K, V>> {
   size_t offset_ = 0;
 };
 
+template <typename V>
+struct ArrayView {
+  using reader_t = VectorReader<V>;
+  using element_t = typename reader_t::exec_in_t;
+
+ public:
+  ArrayView(const reader_t* reader, vector_size_t offset, vector_size_t size)
+      : reader_(reader), offset_(offset), size_(size) {}
+
+  ArrayView() : reader_(nullptr), offset_(0), size_(0) {}
+
+  // Represents an element in the array with an interface to std::optional
+  struct ArrayViewElement {
+    operator bool() const {
+      return this->has_value();
+    }
+
+    bool operator==(const element_t& rhs) const {
+      return has_value() && (rhs == value());
+    }
+
+    inline bool has_value() const {
+      return (*arrayView_.reader_).isSet(arrayView_.offset_ + index_);
+    }
+
+    const element_t value() const {
+      return (*arrayView_.reader_)[arrayView_.offset_ + index_];
+    }
+
+    const element_t operator*() const {
+      return (*arrayView_.reader_)[arrayView_.offset_ + index_];
+    }
+
+   private:
+    ArrayViewElement(const ArrayView& arrayView, size_t index)
+        : arrayView_(arrayView), index_(index) {}
+
+    const ArrayView& arrayView_;
+
+    size_t index_;
+
+    friend class ArrayView;
+  };
+
+  // Iterator class
+  struct IArrayViewElement {
+    FOLLY_ALWAYS_INLINE bool operator!=(const IArrayViewElement& rhs) const {
+      return element_.index_ != rhs.element_.index_;
+    }
+
+    FOLLY_ALWAYS_INLINE const ArrayViewElement& operator*() const {
+      return element_;
+    }
+
+    FOLLY_ALWAYS_INLINE IArrayViewElement& operator++() {
+      ++element_.index_;
+      return *this;
+    }
+
+   private:
+    IArrayViewElement(const ArrayViewElement& element) : element_(element) {}
+
+    ArrayViewElement element_;
+
+    friend class ArrayView;
+  };
+
+  using iterator = IArrayViewElement;
+
+  iterator begin() const {
+    return IArrayViewElement{ArrayViewElement{*this, 0}};
+  }
+
+  iterator end() const {
+    return IArrayViewElement{ArrayViewElement{*this, size_}};
+  }
+
+  // TODO Replace with hasNulls() and check bits in bulk.
```

Comment:
oh what does this return right now? is it returning if the top level vectors may have null and not the array itself ?

Reply:
I am not sure accurate is the mayHaveNulls_ that is set during decoding?
but now its constant time.

Reply:
I see, I see never mind my previous comment, so right now we return the nullability of whole reader.
I will update it.

Reply:
I looked at it, it sounds like we cant bulk check since there is an indirection to get the null index for a specific vector index 
```
    vector_size_t innerIndex = nullIndices_[index];
    return bits::isBitNull(nulls_, innerIndex);
```
I wonder if there is a guarantee that inner indicies will be continuous? if not  
then i think we should keep it as is and the user can iterate over the arrays to check 
nulls if needed., since it will be the same cost.  @mbasmanova  what do you think